### PR TITLE
add Digest Agent

### PR DESCRIPTION
**Pull request by xu-cheng**
Similar to Email Digest Agent, except it will emit a digest event
instead of sending email. It could be useful to send digest message
through meanings other than email.

### CODE REVIEW

**Comment by cantino** [comment]
Nice @xu-cheng, this is a cool idea!  You may want to queue up the event ids instead of the full payloads, then load the events from `received_events.where(ids: memory['event_ids'])` so that the memory in the DB doesn't grow too large?

**Comment by xu-cheng** [comment]
Updated to store event ids. But I have no luck to load events back, would mind to help me to sort this out?

**Comment by cantino** [comment]
Figured it out :)

Add

```ruby
@checker.sources << agents(:bob_rain_notifier_agent)
@checker.sources << agents(:bob_weather_agent)
```

in your "should emit a event" spec so that the Agents are actually wired together.

**Comment by xu-cheng** [comment]
Thanks. Now it works as expected.

File: app/models/agents/digest_agent.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          {
+            "events": [ event list ],
+            "message": "Your message"
+          }
+    MD
+
+    def default_options
+      {
+          "expected_receive_period_in_days" => "2",
+          "message" => "{{ events | join: ',' }}"
```
Comment (cantino) [inline app/models/agents/digest_agent.rb]: Shouldn't this be `events.message | join: ','` or something?

Comment (cantino) [inline app/models/agents/digest_agent.rb]: Or `events | map:'data' | join:';'`

Comment (xu-cheng) [inline app/models/agents/digest_agent.rb]: We don't know what payload are in `events`. So I think this is subjected to user's configuration.

Comment (cantino) [inline app/models/agents/digest_agent.rb]: But `events` here are definitely objects, not strings, right?  I might include something like `events | map:'data' | join:';'` as an example?

Comment (xu-cheng) [inline app/models/agents/digest_agent.rb]: I change it to `events | map: 'message' | join: ';'` since `message` seems to be more common in huginn events.

**Comment by cantino** [comment]
This looks great to me!  The only other question that I have is whether you should add support for sorting the output payloads by `events_order` like the WebsiteAgent, RSSAgent, and DataOutputAgent?  See `description_events_order` and `sort_events`.  I think it might be as simple as:

```ruby
events = sort_events(received_events.where(id: self.memory["queue"]).order(id: :asc).to_a)
```

(Does that look right, @knu?)

**Comment by xu-cheng** [comment]
Sorry about the delay, but may I ask how to use `events_order` in `form_configurable`?

**Comment by cantino** [comment]
Hmm, good question.  It may only work in the JSON editor view?  @knu

**Comment by dsander** [comment]
@cantino I think that is true, `form_configurable` does not a dynamic `options` structure. Could we add support to use the JSON editor only for certain attributes?

**Comment by cantino** [comment]
I was thinking that `form_configurable` mode still showed the optional JSON editor, but I'm remembering incorrectly.  I guess for now we could either make this not `form_configurable`, or not use `events_order` until someone needs it.  I'm fine with either.

**Comment by xu-cheng** [comment]
OK, I kept `form_configurable` and removed `events_order`. Because I found `form_configurable` quite helpful to format final message.

**Comment by cantino** [comment]
Nice @xu-cheng!  Is this done?  I haven't manually tested it, but I trust that you have.  We can merge when you're ready.

**Comment by xu-cheng** [comment]
Yes, it's done.

**Comment by cantino** [comment]
Thanks for contributing this!

**Pull request merged**